File: examples/CBuilder/ExReg1.cpp
//---------------------------------------------------------------------------
#include <vcl.h>
#pragma hdrstop

#include "ExReg1.h"
#include "OnGuard.hpp"
#include "OgUtil.hpp"

//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.dfm"
TfrmExReg *frmExReg;

const TKey EncryptionKey = {0xE5, 0x8F, 0x84, 0xD6, 0x92, 0xC9, 0xA4, 0xD8,
                            0x1A, 0xFA, 0x6F, 0x8D, 0xAB, 0xFC, 0xDF, 0xB4};

//---------------------------------------------------------------------------
__fastcall TfrmExReg::TfrmExReg(TComponent* Owner)
    : TForm(Owner)
{
}
//---------------------------------------------------------------------------
void __fastcall TfrmExReg::btnGenerateClick(TObject *Sender)
{
  TKey Key;
  TCode Code;
  memcpy(Key, EncryptionKey, sizeof(EncryptionKey));
  long Modifier = StringHashElf(edtUserName1->Text);
  ApplyModifierToKeyPrim(Modifier, Key, sizeof(Key));
  InitRegCode(Key, "", pickExpirationDate->Date, Code);
  edtRegCode1->Text = BufferToHex(&Code, sizeof(Code));
}
//---------------------------------------------------------------------------
void __fastcall TfrmExReg::btnVerifyClick(TObject *Sender)
{
  TKey Key;
  TCode Code;
  memcpy(Key, EncryptionKey, sizeof(EncryptionKey));
  long Modifier = StringHashElf(edtUserName1->Text);
  ApplyModifierToKeyPrim(Modifier, Key, sizeof(Key));
  HexToBuffer(edtRegCode2->Text, &Code, sizeof(Code));
  if (IsRegCodeValid(Key, Code))
  {
    if (IsRegCodeExpired(Key, Code))
      edtStatus->Text = "Expired";
    else
      edtStatus->Text = "Valid";
  }
  else
    edtStatus->Text = "Not Valid";
}
//---------------------------------------------------------------------------

File: examples/Delphi/ExReg1.pas
unit ExReg1;

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms, Dialogs,
  StdCtrls, ComCtrls, Buttons;

type
  TfrmExRegCode = class(TForm)
    GroupBox1: TGroupBox;
    Label1: TLabel;
    edtUserName1: TEdit;
    Label2: TLabel;
    pickExpirationDate: TDateTimePicker;
    Label3: TLabel;
    edtRegCode1: TEdit;
    GroupBox2: TGroupBox;
    edtUserName2: TEdit;
    Label4: TLabel;
    edtRegCode2: TEdit;
    Label5: TLabel;
    Label6: TLabel;
    edtStatus: TEdit;
    btnGenerate: TSpeedButton;
    btnVerify: TSpeedButton;
    procedure btnGenerateClick(Sender: TObject);
    procedure btnVerifyClick(Sender: TObject);
  private
    { Private declarations }
  public
    { Public declarations }
  end;

var
  frmExRegCode: TfrmExRegCode;

implementation

{$R *.DFM}

uses
  OnGuard, OgUtil;

const
  EncryptionKey : TKey = ($E5, $8F, $84, $D6, $92, $C9, $A4, $D8,
                          $1A, $FA, $6F, $8D, $AB, $FC, $DF, $B4);


procedure TfrmExRegCode.btnGenerateClick(Sender: TObject);
var
  Key : TKey;
  Code : TCode;
  Modifier : Longint;
begin
  Key := EncryptionKey;
  Modifier := StringHashELF(edtUserName1.Text);
  ApplyModifierToKeyPrim(Modifier, Key, SizeOf(Key));
  InitRegCode(Key, '', pickExpirationDate.Date, Code);
  edtRegCode1.Text := BufferToHex(Code, SizeOf(Code));
end;

procedure TfrmExRegCode.btnVerifyClick(Sender: TObject);
var
  Key : TKey;
  Code : TCode;
  Modifier : Longint;
begin
  Key := EncryptionKey;
  Modifier := StringHashELF(edtUserName2.Text);
  ApplyModifierToKeyPrim(Modifier, Key, SizeOf(Key));
  HexToBuffer(edtRegCode2.Text, Code, SizeOf(Code));
  if IsRegCodeValid(Key, Code) then begin
    if IsRegCodeExpired(Key, Code) then
      edtStatus.Text := 'Expired'
    else
      edtStatus.Text := 'Valid';
  end else
    edtStatus.Text := 'Not Valid';
end;

end.
